File: Practise/practice/MergeSort.cpp
#include <iostream>
#include <vector>

using namespace std;

void merge(vector<int> &arr, int low, int mid, int high);

void mS(vector<int> &arr, int low, int high)
{
    if (low == high) return;
    int mid = (low + high) / 2;
    mS(arr, low, mid);
    mS(arr, mid + 1, high);
    merge(arr, low, mid, high);
}

void mergeSort(vector<int> &arr, int n)
{
    mS(arr, 0, n - 1);
}

void merge(vector<int> &arr, int low, int mid, int high)
{
    vector<int> temp;
    int left = low;
    int right = mid + 1;

    while (left <= mid && right <= high)
    {
        if (arr[left] <= arr[right])
        {
            temp.push_back(arr[left]);
            left++;
        }
        else
        {
            temp.push_back(arr[right]);
            right++;
        }
    }

    while (left <= mid)
    {
        temp.push_back(arr[left]);
        left++;
    }

    while (right <= high)
    {
        temp.push_back(arr[right]);
        right++;
    }

    for (int i = low; i <= high; i++)
    {
        arr[i] = temp[i - low];
    }
}

int main()
{
    // int arr[] = {3, 1, 2, 4, 1, 5, 2, 6, 4};
    vector<int> arr = {3, 2, 4, 1, 3};  // Use vector instead of an array

    mergeSort(arr, arr.size() - 1);  // Pass the size of the vector

    for (int i = 0; i < arr.size(); i++)
    {
        cout << arr[i] << "\t";
    }

    return 0;
}

File: DSA/Sorting/MergeSort.java
import java.util.ArrayList;
import java.util.List;

public class MergeSort {

    public static void mergeSort(int[] array, int low, int high) {
        // Base Case
        if (low >= high) {
            return;
        }

        int mid = (low + high) / 2;

        // Left half
        mergeSort(array, low, mid);

        // Right half
        mergeSort(array, mid + 1, high);

        // Merge the two halves
        merge(array, low, mid, high);
    }

    public static void merge(int array[], int low, int mid, int high) {
        List<Integer> temp = new ArrayList<>(); // Temporary array
        int left = low;      // Starting index of the left half of array
        int right = mid + 1; // Starting index of the right half of array

        // Sorting elements into the temporary array in a sorted manner
        while (left <= mid && right <= high) {
            if (array[left] <= array[right]) {
                temp.add(array[left]);
                left++;
            } else {
                temp.add(array[right]);
                right++;
            }
        }

        // If elements on the left half are still left
        while (left <= mid) {
            temp.add(array[left]);
            left++;
        }

        // If elements on the right half are still left
        while (right <= high) {
            temp.add(array[right]);
            right++;
        }

        // Copy the sorted elements back into the original array
        for (int i = low; i <= high; i++) {
            array[i] = temp.get(i - low);
        }
    }

    public static void main(String[] args) {
        int[] array = {38, 27, 43, 3, 9, 82, 10};
        int low = 0;
        int high = array.length - 1;
        mergeSort(array, low, high);

        // Print the sorted array
        for (int num : array) {
            System.out.print(num + " ");
        }
    }
}
